File: ruby/array_hashing/49_group_anagrams.rb
# @param {String[]} strs
# @return {String[][]}
def group_anagrams(strs)
  return [strs] if strs.size <= 1

  result = Hash.new([])
  count = Array.new(26, 0)

  strs.each do |str|
    str.each_char { |char| count[char.ord - "a".ord] += 1 }
    if result.key?(count)
      result[count].push(str)
    else
      result[count] = [str]
    end
    count = Array.new(26, 0)
  end

  result.values
end

def test
  # Test cases

  # result: [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]]
  puts group_anagrams(%w[eat tea tan ate nat bat]).inspect

  # result: [[""]]
  puts group_anagrams([""]).inspect

  # result: [["a"]]
  puts group_anagrams(["a"]).inspect

  # result: [["a", "a"], ["aa"], ["ab"]]
  puts group_anagrams(%w[a a aa ab]).inspect
end

test

File: go/array_hashing/49_group_anagrams.go
package main

import "fmt"

func main() {
	// Test cases

	// result: [["bat"],["nat","tan"],["ate","eat","tea"]]
	fmt.Println(groupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))

	// result: [[]]
	fmt.Println(groupAnagrams([]string{""}))

	// result: [["a"]]
	fmt.Println(groupAnagrams([]string{"a"}))

	// result: [["a", "a"], ["aa"], ["ab"]]
	fmt.Println(groupAnagrams([]string{"a", "a", "aa", "ab"}))
}

func groupAnagrams(strs []string) [][]string {
	groups := make(map[[26]int][]string)

	for _, str := range strs {
		group := [26]int{}
		for _, char := range []rune(str) {
			group[char-'a'] += 1
		}
		if _, exists := groups[group]; exists {
			groups[group] = append(groups[group], str)
		} else {
			groups[group] = []string{str}
		}
	}

	result := make([][]string, 0, len(groups))

	for _, group := range groups {
		result = append(result, group)
	}

	return result
}
